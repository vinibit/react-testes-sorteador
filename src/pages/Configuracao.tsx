import React from "react";
import Cadastro from "../components/Participante/Cadastro/Cadastro";
import Lista from "../components/Participante/Lista/Lista";
import Footer from "../components/Footer/Footer";
import Card from "../components/Card/Card";

const Configuracao: React.FC = () => {

    return (
        <Card>
            <section>
                <h2>Vamos come√ßar!</h2>
                <Cadastro />
                <Lista />
                <Footer />
            </section>
        </Card>
    )
}

export default Configuracao