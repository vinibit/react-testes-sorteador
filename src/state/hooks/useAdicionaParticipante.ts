import { useRecoilValue, useSetRecoilState } from "recoil"
import { listaParticipantesState, erroState } from "../atom"
import Participante from "../../types/Participante"

export const useAdicionaParticipante = () => {

    const setListaParticipantes = useSetRecoilState(listaParticipantesState)
    const listaAtual = useRecoilValue(listaParticipantesState)
    const setErro = useSetRecoilState(erroState)

    return (nome: string) => {
        
        if (listaAtual.some(participante => participante.nome === nome)) {
            setErro("Participante jÃ¡ adicionado na lista!")
            setTimeout(() => {
                setErro("")
            }, 5000)
            return
        }

        const id = Math.random().toString(36).substring(2, 9)
        const novoParticipante: Participante = { id, nome }
        setListaParticipantes(listaAtual => [ ...listaAtual, novoParticipante])
    }    
}