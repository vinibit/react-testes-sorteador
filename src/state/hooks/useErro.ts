import { useRecoilValue } from "recoil"
import { erroState } from "../atom"

export const useErro = () => {
    const erro = useRecoilValue(erroState)
    return erro
}