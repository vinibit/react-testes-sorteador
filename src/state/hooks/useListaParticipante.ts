import { useRecoilValue } from "recoil"
import { listaParticipantesState } from "../atom"

export const useListaParticipante = () => {
    return useRecoilValue(listaParticipantesState)
}