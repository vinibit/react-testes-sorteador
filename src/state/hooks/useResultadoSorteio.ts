import { useRecoilValue } from 'recoil'
import { resultadoSorteioState } from '../atom'

export const useResultadoSorteio = () => {
    return useRecoilValue(resultadoSorteioState)
}