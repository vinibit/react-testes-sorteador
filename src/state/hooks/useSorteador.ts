
import { useSetRecoilState } from "recoil"
import { resultadoSorteioState } from "../atom"
import { useListaParticipante } from "./useListaParticipante"
import { sortear } from "../helpers/sortear"

export const useSorteador = () => {

    const setResultado = useSetRecoilState(resultadoSorteioState)    
    const participantes = useListaParticipante()

    return () => {
        const resultado = sortear(participantes)        
        setResultado(resultado)
    }
}