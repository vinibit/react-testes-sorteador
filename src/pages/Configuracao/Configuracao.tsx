import styles from "./Configuracao.module.css";

import React from "react";
import { useNavigate } from "react-router";
import Cadastro from "../../components/Participante/Cadastro/Cadastro";
import Lista from "../../components/Participante/Lista/Lista";
import Footer from "../../components/Footer/Footer";
import Card from "../../components/Card/Card";
import { useSorteador } from "../../state/hooks/useSorteador";
import { useListaParticipante } from "../../state/hooks/useListaParticipante";

const Configuracao: React.FC = () => {

    const navigate = useNavigate()
    const sortear = useSorteador()
    const participantes = useListaParticipante()
    const temParticipantesSuficientes = participantes.length >= 3

    const iniciar = () => {
        if (!temParticipantesSuficientes) {
            alert("É necessário ter pelo menos 3 participantes para iniciar a brincadeira.")
            return
        }

        sortear()
        navigate("/sorteio")        
    }

    return (
        <Card>
            <section>
                <h2>Vamos começar!</h2>
                <Cadastro />
                <Lista />
                <Footer>
                    <button 
                        type="button"
                        data-testid="botao-iniciar"
                        className={styles.botaoIniciar}
                        disabled={!temParticipantesSuficientes} 
                        onClick={iniciar}>                    
                            <span className="material-icons">play_circle_outline</span>
                            Iniciar brincadeira
                    </button>
                    <img src="/imagens/sacolas.png" alt="Sacolas de compras" />                
                </Footer>
            </section>
        </Card>
    )
}

export default Configuracao