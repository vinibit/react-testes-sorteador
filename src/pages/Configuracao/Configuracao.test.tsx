import { fireEvent, render, screen } from "@testing-library/react";
import { RecoilRoot } from 'recoil';
import { useListaParticipante } from "../../state/hooks/useListaParticipante";
import Configuracao from './Configuracao';
import { useSorteador } from "../../state/hooks/useSorteador";

const component = (
    <RecoilRoot>        
        <Configuracao />
    </RecoilRoot>    
)

const mockNavigate = jest.fn()
const mockSorteio = jest.fn()

jest.mock("react-router", () => {
    return {
        useNavigate: () => mockNavigate
    }
})

jest.mock("../../state/hooks/useListaParticipante", () => {
    return {
        useListaParticipante: jest.fn()
    }
})

jest.mock("../../state/hooks/useSorteador", () => {
    return {
        useSorteador: jest.fn()
    }
})

describe("Na página de Configuração", () => {

    beforeEach(() => {
        (useListaParticipante as jest.Mock).mockReturnValue([])
    })
    
    test("Deve ser renderizada corretamente", () => {
        
        const { container } = render(component);
        expect(container).toMatchSnapshot();
    })

})


describe("Quando não existirem participantes suficientes", () => {

    beforeEach(() => {
        (useListaParticipante as jest.Mock).mockReturnValue([])
    })
    
    test("Deve renderizar o Rodapé", () => {
        
        render(component)

        const botao = screen.getByTestId("botao-iniciar")
        expect(botao).toBeDisabled()
    })

})

describe("Quando existirem participantes suficientes", () => {

    const participantes = [
        { id: 1, nome: "Participante 1" },
        { id: 2, nome: "Participante 2" },
        { id: 3, nome: "Participante 3" },
    ]

    beforeEach(() => {
        (useListaParticipante as jest.Mock).mockReturnValue(participantes);
        (useSorteador as jest.Mock).mockReturnValue(mockSorteio)
    })
    
    test("Pode iniciar a brincadeira", () => {
        
        render(component)

        const botao = screen.getByTestId("botao-iniciar")
        expect(botao).not.toBeDisabled()
    })

    test("A brincadeira foi iniciada", () => {
        
        render(component)

        const botao = screen.getByTestId("botao-iniciar")
        fireEvent.click(botao)

        expect(mockNavigate).toHaveBeenCalledTimes(1)
        expect(mockNavigate).toHaveBeenCalledWith("/sorteio")
        expect(mockSorteio).toHaveBeenCalledTimes(1)
    })
})