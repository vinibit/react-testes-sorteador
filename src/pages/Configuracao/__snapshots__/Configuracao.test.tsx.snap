// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Na página de Configuração Deve ser renderizada corretamente 1`] = `
<div>
  <div
    class="card"
  >
    <section>
      <h2>
        Vamos começar!
      </h2>
      <form>
        <div
          class="grupoInputBotaoCadastro"
        >
          <input
            class="inputNome"
            placeholder="Insira o nome dos participantes"
            type="text"
            value=""
          />
          <button
            data-testid="botao-adicionar"
            disabled=""
            type="submit"
          >
            Adicionar
          </button>
        </div>
      </form>
      <ul
        role="list"
      />
      <footer
        class="footer"
      >
        <button
          class="botaoIniciar"
          data-testid="botao-iniciar"
          disabled=""
          type="button"
        >
          <span
            class="material-icons"
          >
            play_circle_outline
          </span>
          Iniciar brincadeira
        </button>
        <img
          alt="Sacolas de compras"
          src="/imagens/sacolas.png"
        />
      </footer>
    </section>
  </div>
</div>
`;
